### DISCUSSION TITLE
GPX TBT navigation track (offline)

### DISCUSSION THREAD

**Discussion by 0709wiwiwi** [devemux86/cruiser#250] [Show and tell]
Gpx navigation track by a gpx 1.1 conform file.

The gpx navigation track mode exclusively used and generated in Cruiser (importable in Locus) is still hardly known.  
Therefore, I share in the attached pdf how simple a gpx navigation track is put together.
Even with Notepad ++ you can modify or tune a navigation track if you should want to.

[Composition of a gpx navigation track.pdf](https://github.com/devemux86/cruiser/files/12379328/Composition.of.a.gpx.navigation.track.pdf)

### Expressions in gpx.

**- (wpt) waypoint.**   (_waypoints are **SOLO** players_) 
   _Indicates the location of a **single** point of interest._
   _The waypoint locations are not part of a route for navigation._
   _The waypoint order in the waypoint list is so not important._
  
  But anyway they may and can be added into a route as routepoints subsequently . 
  Strictly ordered waypoints added in a **route** are so promoted into **route**(way)**points**.

**- (rte) route with (rtept) routepoints.**  (_routepoints are group players_)
   _A route contains multiple consecutive via or shaping routepoints.
   The routepoints are visually connected by straight lines on the map.
   These lines can be partially or even completely be transparent.
    In navigation the via variant is announced while the shaping variant is muted.
   The routepoints are those locations you definitely want to pass **nearby**.
   The Router engine calculates a track by which roads the locations are reached._

**- (trk) track with (trkpt) trackpoints.** (_trackpoints are mass players_)
   _A track contains multiple trackpoints calculated by the Router.
   A track to follow is accurately trackable due to its high trackpoint density._

**Extra. (Cruiser)**

A navigation track with navigation trackpoints. (navtrk by navtrkpt)
An **enriched** track with shaping & via trackpoints plus the turn instructions.
The shaping or via point **name** is attached in the trackpoint **name**.
The **streetname** is attached in the trackpoint **cmt**.  Unknown streetname is represented by "--"
Extra info can be attached into a via trackpoint **desc**.  Tap the via point icon shows  the extra.
_Via points Icons in the Cruiser app are by default yellow colored pins.
Marking those via points that do contain more info in **desc** is possible.
This by assigning a different (free) **color** in the gpx trackpoint extension_ 
The tracksection routing **profile** is attached in the target (section-end) via(shaping) trackpoint **src**.
The **turn** instruction is attached in the trackpoint **sym**.
The **shaping** (silent) or **via** (announced) point is attached in the trackpoint **type**.

_A trackpoint sym (Locus) promotes the location into a turn instruction point.
A trackpoint type (Cruiser) is accompanied by a trackpoint sym.
A trackpoint type indicates the "snapped to street" routepoint locations.
By a trackpoint type **via** this point is represented by the **via** point icon.
A trackpoint type **shaping** is represented by the **shaping** point icon._

**Shaping & Via points.**

_Cruiser adds a via or shaping marker in the type element of as well the **routepoints** and **trackpoints**. 
Garmin uses a specific Garmin gpx extension to add a via or shaping marker in the **routepoints** only._

**Comment by devemux86** [comment]
@0709wiwiwi thanks for the document!

BRouter also now uses the gpx navigation track format in the calculated route responses.
@0709wiwiwi and I worked with @afischerdev testing BRouter during implementation.

---

[BRouter wesbite](https://brouter.de/brouter-web/) provides the gpx navigation track format in the export:
- Select **locus-style** in **turnInstructionMode**
- **Export route** pressing the **cloud** button

So we can design routes on BRouter website, import the gpx **offline** in Cruiser and start navigation!

![BRouter web client](https://github.com/devemux86/cruiser/assets/3484020/f5ea6c19-aae7-4990-b0f4-a4b84b6753da)

**Reply by 0709wiwiwi** [reply]
An interesting Brouter web version for cyclists is [Brouter web essbee]( https://brouter.de/essbee/#map=14/51.5952/5.6909/standard,route-quality)
**Use**
In the profile options, select: locus-style plus set the other personal preferences as desired. 
Important **PUSH APPLY!**
Save a design by the Export route function. Done.

**Reply by devemux86** [reply]
@0709wiwiwi Thanks, the BRouter websites are mentioned here:
- https://github.com/devemux86/cruiser/discussions/272

**Comment by 0709wiwiwi** [comment]
The gpx navigation track contains the basic functions without gpx extensions.
The less necessary extras (color) are eventually to be attached in a gpx extension.

[Demo.attach.gpx.extensions.gpx.txt](https://github.com/user-attachments/files/25458763/Demo.attach.gpx.extensions.gpx.txt)

<img width="788" height="583" alt="Demo attach gpx extensions" src="https://github.com/user-attachments/assets/15c8142a-fc05-47cc-a088-e89d93a078f2" />


[Demo.attach.extensions.gpx.metadata.rte.rtept.trk.trkpt.gpx.txt](https://github.com/user-attachments/files/25458779/Demo.attach.extensions.gpx.metadata.rte.rtept.trk.trkpt.gpx.txt)

**Reply by 0709wiwiwi** [reply]
In valid gpx the element order is very strict as described in the standard.
In the gpx extension applied in Cruiser the extra element order is free.
A Cruiser gpx file contains no extra private xml namespace definition.
Extra elements use the xsi namespace as no schema “cuckoo” extension.  

Extra elements in Cruiser gpx.
- color #AARRGGBB or #RRGGBB (track and gpx wpt icon color)
- width (trackline width)
- speed (m/s)
- max_speed (m/s)
- route_information (paved - unpaved)
